File: backend/app/com/skewwhiffy/auraltester/services/ScaleService.scala
package com.skewwhiffy.auraltester.services

import com.skewwhiffy.auraltester.clefs.Clef
import com.skewwhiffy.auraltester.notes.Note
import com.skewwhiffy.auraltester.scales.{Scale, ScaleDirection, ScaleType}

import scala.util.chaining.scalaUtilChainingOps

class ScaleService {
  def getScale(clef: Clef, note: Note, scaleType: ScaleType, direction: ScaleDirection): Scale = {
    clef.getNoteNearBottom(note).pipe(it => new Scale(it, scaleType, direction))
  }
}

File: backend/src/main/kotlin/com/skewwhiffy/service/ScaleService.kt
package com.skewwhiffy.service

import com.skewwhiffy.notation.model.clef.Clef
import com.skewwhiffy.notation.model.note.Note
import com.skewwhiffy.notation.model.scale.Scale
import com.skewwhiffy.notation.model.scale.ScaleDirection
import com.skewwhiffy.notation.model.scale.ScaleType
import org.springframework.stereotype.Service

@Service
class ScaleService {
  fun getScale(
      clef: Clef,
      note: Note,
      scaleType: ScaleType,
      direction: ScaleDirection,
  ): Scale {
    return clef.getNoteNearBottom(note).let { Scale(it, scaleType, direction) }
  }
}